### PR TITLE
Split UniformHolder interface

### PR DESCRIPTION
**Pull request by Kroppeb**
(Needed for Custom Uniforms)

This pr moves part of the code of `UniformHolder` into `LocationalUniformHolder` as the Custom Uniform input manager can't return any useful `location` nor can it accept arbitrary uniforms.

### CODE REVIEW

File: src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java [hunk 1]
```diff
@@ -32,7 +33,7 @@ private SamplerUniforms() {
 		// no construction allowed
 	}
 
-	public static void addCommonSamplerUniforms(UniformHolder uniforms) {
+	public static void addCommonSamplerUniforms(LocationalUniformHolder uniforms) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java:36]: Does this need to be a locational uniform holder, feels like UniformHolder work be acceptable here too?

File: src/main/java/net/coderbot/iris/uniforms/CommonUniforms.java [hunk 2]
```diff
@@ -31,14 +26,17 @@
...
+
 public final class CommonUniforms {
 	private static final MinecraftClient client = MinecraftClient.getInstance();
 
 	private CommonUniforms() {
 		// no construction allowed
 	}
 
-	public static void addCommonUniforms(UniformHolder uniforms, IdMap idMap) {
+	public static void addCommonUniforms(LocationalUniformHolder uniforms, IdMap idMap) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/uniforms/CommonUniforms.java:39]: Does this need to be locational?

File: src/main/java/net/coderbot/iris/gl/uniform/Uniform.java [hunk 3]
```diff
@@ -3,7 +3,7 @@
 public abstract class Uniform {
 	protected final int location;
 
-	Uniform(int location) {
+	public Uniform(int location) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/uniform/Uniform.java:6]: Why does this need to be public?

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Looks mostly good, I have a few comments though. Also, this PR currently conflicts with trunk it appears.

File: src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java (hunk 1 above)
Comment (Kroppeb) [reply src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java:36]: The samplers need the `location` to select the correct `shadow` sampler. It changes depending on whether the `watershadow` is used tr not it seems.

File: src/main/java/net/coderbot/iris/uniforms/CommonUniforms.java (hunk 2 above)
Comment (Kroppeb) [reply src/main/java/net/coderbot/iris/uniforms/CommonUniforms.java:39]: yep, see https://github.com/IrisShaders/Iris/pull/238#discussion_r611900460

File: src/main/java/net/coderbot/iris/gl/uniform/Uniform.java (hunk 3 above)
Comment (Kroppeb) [reply src/main/java/net/coderbot/iris/gl/uniform/Uniform.java:6]: Oh, this is shouldn't be in this commit. I think in one of my latest branches this doesn't even need to be public anymore. 

I'll remove it.

File: src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java (hunk 1 above)
Comment (coderbot16) [reply src/main/java/net/coderbot/iris/uniforms/SamplerUniforms.java:36]: Right, forgot about that, in that case this is fine for now. Might be worth adding comments about that though.

**Review by coderbot16 — APPROVED** [review]
Thanks!

**Pull request merged**